spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/observation/ObservedAppDeployer.java

```diff
@@ -0,0 +1,227 @@
+			registerListener(observation, id);
+			return id;
+		});
+	}
+
+	@Override
+	public AppStatus status(String id) {
+		AppDeployerContext context = new AppDeployerContext(environmentInfo());
+		context.setAppId(id);
+		Observation observation = AppDeployerDocumentedObservation.DEPLOYER_STATUS_OBSERVATION.start(this.customConvention, DEFAULT_CONVENTION, context, this.observationRegistry);
+		return observation.observe(() -> this.delegate.status(id));
+	}
+
+	@Override
+	public Mono<AppStatus> statusReactive(String id) {
+		return this.delegate.statusReactive(id);
+		// TODO: Fix me
+//		return ReactorSleuth.tracedMono(tracer(), currentTraceContext(),
+//				SleuthDeployerSpan.DEPLOYER_STATUS_SPAN.getName(), () -> this.delegate.statusReactive(id),
+//				(o, span) -> span.tag(SleuthDeployerSpan.Tags.APP_ID.getKey(), id),
+//				span -> clientSpan(SleuthDeployerSpan.DEPLOYER_STATUS_SPAN, span).start());
+	}
+
+	@Override
+	public Flux<AppStatus> statusesReactive(String... ids) {
+		return this.delegate.statusesReactive(ids);
+		// TODO: Fix me
+//		return ReactorSleuth.tracedFlux(tracer(), currentTraceContext(),
+//				SleuthDeployerSpan.DEPLOYER_STATUSES_SPAN.getName(), () -> this.delegate.statusesReactive(ids),
+//				(o, span) -> span.tag(SleuthDeployerSpan.Tags.APP_ID.getKey(), Arrays.toString(ids)),
+//				span -> clientSpan(SleuthDeployerSpan.DEPLOYER_STATUSES_SPAN, span).start());
+	}
+
+	@Override
+	public RuntimeEnvironmentInfo environmentInfo() {
+		return this.delegate.environmentInfo();
+	}
+
+	@Override
+	public String getLog(String id) {
+		AppDeployerContext context = new AppDeployerContext(environmentInfo());
+		context.setAppId(id);
+		Observation observation = AppDeployerDocumentedObservation.DEPLOYER_GET_LOG_OBSERVATION.start(this.customConvention, DEFAULT_CONVENTION, context, this.observationRegistry);
+		return observation.observe(() -> this.delegate.getLog(id));
+	}
+
+	@Override
+	public void scale(AppScaleRequest appScaleRequest) {
+		AppDeployerContext context = new AppDeployerContext(environmentInfo());
+		context.setAppScaleRequest(appScaleRequest);
+		Observation observation = AppDeployerDocumentedObservation.DEPLOYER_SCALE_OBSERVATION.start(this.customConvention, DEFAULT_CONVENTION, context, this.observationRegistry);
+		observation.observe(() -> this.delegate.scale(appScaleRequest));
+	}
+
+	private static final class PreviousAndCurrentStatus {
+
+		private final Observation observation;
+
+		private AppStatus current;
+
+		private AppStatus previous;
+
+		private PreviousAndCurrentStatus(Observation observation) {
+			this.observation = observation;
+			if (log.isDebugEnabled()) {
+				log.debug("Current observation is [" + observation + "]");
+			}
+		}
+
+		private PreviousAndCurrentStatus updateCurrent(AppStatus current) {
+			if (log.isTraceEnabled()) {
+				log.trace("State before change: current [" + this.current + "], previous [" + this.previous + "]");
+			}
+			this.previous = this.current;
```

Comment:
Any thread-safety concerns here? If so, and we only care about the previous to register a status changed or not, we could leverage AtomicReference using AppStatus as the ref.

Reply:
I don't think so - here we have statuses changing for a single id.